sdk/examples/poet_python/sawtooth_poet/engine.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2018 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+import logging
+import queue
+import time
+
+from sawtooth_sdk.consensus.engine import Engine
+from sawtooth_sdk.consensus import exceptions
+from sawtooth_sdk.protobuf.validator_pb2 import Message
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class PoetEngine(Engine):
+    def __init__(self):
+        self._exit = False
+        self._service = None
+        self._chain_head = None
+
+    def name(self):
+        return 'PoET'
+
+    def version(self):
+        return '0.1'
+
+    def stop(self):
+        self._exit = True
+
+    def _initialize_block(self):
+        self._service.initialize_block(None)
+
+    def _check_consensus(self, block):
+        return True
+
+    def _compare_forks(self, current_head, new_head):
+        return True
+
+    def _check_block(self, block_id):
+        self._service.check_blocks([block_id])
+
+    def _fail_block(self, block_id):
+        self._service.fail_block(block_id)
+
+    def _get_chain_head(self):
+        return self._service.get_chain_head()
+
+    def _get_block(self, block_id):
+        return self._service.get_blocks([block_id])
+
+    def _commit_block(self, block_id):
+        self._service.commit_block(block_id)
+
+    def _ignore_block(self, block_id):
+        self._service.ignore_block(block_id)
+
+    def _cancel_block(self):
+        try:
+            self._service.cancel_block()
+        except exceptions.InvalidState:
+            pass
+
+    def _finalize_block(self):
+        time.sleep(1)
+
+        while True:
+            try:
+                block_id = self._service.finalize_block(b'consensus')
+                break
+            except exceptions.BlockNotReady:
+                time.sleep(1)
+                continue
+            except exceptions.InvalidState:
+                return None
+
+        return block_id
+
+    def _check_publish_block(self):
+        return True
+
+    def start(self, updates, service, chain_head, peers):
```

Comment:
What is `service`? Please add a doc string on these params.

Reply:
`Service` is a component analogous to a transaction processor's `Context`. Engines use services to make calls to the validator through methods like `commit_block`, `broadcast`, etc. Docstrings are available from the abstract base classes:

`Engine`: https://github.com/hyperledger/sawtooth-core/blob/master/sdk/python/sawtooth_sdk/consensus/engine.py
`Service`: https://github.com/hyperledger/sawtooth-core/blob/master/sdk/python/sawtooth_sdk/consensus/service.py

Reply:
Thanks for the clarifications and links.
Which component will be responsible for calling start? (I don't think we have any design docs on this yet)

Reply:
`Driver` is analogous to `TransactionProcessor` and `Engine` is analogous to `TransactionHandler`. When the driver is started in `main` (https://github.com/nick-drozd/sawtooth-core/blob/4d4181115d139b6547a29568adac030f7bfe8a31/sdk/examples/poet_python/sawtooth_poet/main.py#L73), it will in turn start the engine.